osutil/user/user.go

```diff
@@ -0,0 +1,281 @@
+	return u.home
+}
+
+func (u *User) UID() sys.UserID {
+	return u.uid
+}
+
+func (u *User) GID() sys.GroupID {
+	return u.gid
+}
+
+func (u *User) isSystemUser() bool {
+	if u.uid > 0 && u.uid < minUserUID {
+		return true
+	}
+
+	if u.name == "nobody" {
+		return true
+	}
+
+	return !strutil.ListContains(shells, u.shell)
+}
+
+var me *User
+var meErr error
+
+func Current() (*User, error) {
+	return me, meErr
+}
+
+// NotFound is returned by First when no users are found that match the given filter.
+var NotFound = errors.New("user not found")
+
+// First tries to return a user that matches the given filter.
+func First(filter func(*User) bool) (*User, error) {
+	var it Iter
+	defer it.Finish()
+	for it.Scan(filter) {
+		return it.User(), nil
+	}
+
+	err := NotFound
+	if it.err != nil {
+		err = it.err
+	}
+
+	return nil, err
+}
+
+// FromUID tries to find a user with the given UID.
+func FromUID(uid sys.UserID) (*User, error) {
+	return First(func(u *User) bool {
+		return u.uid == uid
+	})
+}
+
+// FromName tries to find a user with the given name.
+func FromName(name string) (*User, error) {
+	return First(func(u *User) bool {
+		return u.name == name
+	})
+}
+
+var passwds = []string{
+	// order could be important
+	"/var/lib/extrausers/passwd",
+	"/etc/passwd",
+}
+
+// an Iter will iterate over the user database.
+type Iter struct {
+	pwi     int
+	pwf     *os.File
+	scanner *bufio.Scanner
+	cur     *User
+	err     error
+}
+
+// User returns the user found by Scan.
+func (it *Iter) User() *User {
+	return it.cur
+}
+
+// Scan advances the iterator until it finds a user that matches all
+// the given conditions.
+func (it *Iter) Scan(conds ...func(*User) bool) bool {
+	if it.err != nil {
+		return false
+	}
+
+	for it.scanner == nil {
+		if it.pwi >= len(passwds) {
+			// no more files to scan
+			return false
+		}
+		if it.pwf != nil {
+			it.err = it.pwf.Close()
+		}
+		if it.err != nil {
+			return false
+		}
+		it.pwf, it.err = os.Open(filepath.Join(dirs.GlobalRootDir, passwds[it.pwi]))
+		it.pwi++
+		if it.err != nil {
+			// ignore missing files
+			if !os.IsNotExist(it.err) {
+				return false
+			}
+			it.err = nil
+			// try next file
+			continue
+		}
+		it.scanner = bufio.NewScanner(it.pwf)
+	}
+
+outer:
+	for it.scanner.Scan() {
+		it.cur = nil
+		line := bytes.TrimSpace(it.scanner.Bytes())
+		if len(line) == 0 || line[0] == '#' {
+			// blank or comment; ignore
+			continue
+		}
+		fields := bytes.SplitN(line, []byte{':'}, 8)
+		if len(fields) != 7 {
+			// bogus line; ignore
+			continue
+		}
+		// root:x:0:0:root:/root:/bin/bash
+		uid, err := strconv.ParseUint(string(fields[2]), 10, 32)
+		if err != nil {
+			continue
+		}
+		gid, err := strconv.ParseUint(string(fields[3]), 10, 32)
+		if err != nil {
+			continue
+		}
+		u := &User{
+			name:  string(fields[0]),
+			home:  string(fields[5]),
+			shell: string(fields[6]),
+			uid:   sys.UserID(uid),
+			gid:   sys.GroupID(gid),
+		}
+		for _, cond := range conds {
```

Comment:
If this becomes a helper, the extra `outer` could be avoided, maybe something like:
```
        all := func(conds ...func(*User) bool) bool {
		for _u, cond := range conds {
			if !cond(u) {
				return false
			}
		}
		return true
	}
	if !all(conds) {
		continue
	}
```
(or move all out of this scope).